File: temporal/worker-ts/src/workflows/childFanout.ts
import { executeChild, log } from '@temporalio/workflow';
import { proxyActivities } from '@temporalio/workflow';
import type * as activities from '../activities/childActivities';

const { processItem } = proxyActivities<typeof activities>({
  startToCloseTimeout: '30s',
  retry: { maximumAttempts: 2 },
});

export interface ChildFanoutInput {
  items: string[];
  concurrencyLimit?: number; // defaults to 3
}

export interface ChildItemResult {
  itemId: string;
  status: string;
  processedBy: string;
}

export interface ChildFanoutResult {
  totalItems: number;
  completed: number;
  failed: number;
  results: ChildItemResult[];
}

/**
 * Child Fanout Workflow — TypeScript Variant
 *
 * DIFFERENCE FROM GO: Adds a CONCURRENCY LIMIT. Instead of launching all
 * children at once, this processes items in batches to demonstrate
 * controlled parallelism. Default concurrency: 3.
 */
export async function childFanoutWorkflow(input: ChildFanoutInput): Promise<ChildFanoutResult> {
  const concurrencyLimit = input.concurrencyLimit ?? 3;
  log.info('childFanoutWorkflow started', {
    itemCount: input.items.length,
    concurrencyLimit,
  });

  const results: ChildItemResult[] = [];
  let completed = 0;
  let failed = 0;

  // Process items in batches of `concurrencyLimit`
  for (let i = 0; i < input.items.length; i += concurrencyLimit) {
    const batch = input.items.slice(i, i + concurrencyLimit);
    const batchNum = Math.floor(i / concurrencyLimit) + 1;
    log.info(`Processing batch ${batchNum}`, { items: batch });

    // Launch child workflows for this batch in parallel
    const batchPromises = batch.map((item) =>
      executeChild(childProcessItemWorkflow, {
        args: [item],
        workflowId: `child-ts-${item}-${Date.now()}`,
      }).catch((err: Error) => {
        log.warn('Child workflow failed', { item, error: err.message });
        return {
          itemId: item,
          status: `failed: ${err.message}`,
          processedBy: 'ts-worker',
        } as ChildItemResult;
      })
    );

    // Wait for all in this batch to complete before starting next batch
    const batchResults = await Promise.all(batchPromises);

    for (const result of batchResults) {
      results.push(result);
      if (result.status === 'completed') {
        completed++;
      } else {
        failed++;
      }
    }

    log.info(`Batch ${batchNum} complete`, {
      batchCompleted: batchResults.filter((r) => r.status === 'completed').length,
    });
  }

  log.info('childFanoutWorkflow completed', {
    total: input.items.length,
    completed,
    failed,
  });

  return {
    totalItems: input.items.length,
    completed,
    failed,
    results,
  };
}

/**
 * Child Process Item Workflow — processes a single item.
 * Called by childFanoutWorkflow as a child workflow.
 */
export async function childProcessItemWorkflow(itemId: string): Promise<ChildItemResult> {
  log.info('childProcessItemWorkflow started', { itemId });

  const result = await processItem(itemId);

  return {
    itemId,
    status: 'completed',
    processedBy: 'ts-worker',
  };
}

File: temporal/worker-go/workflows/child_fanout.go
package workflows

import (
	"fmt"
	"time"

	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"

	"temporal-worker-go/activities"
)

// ChildFanoutInput contains items to process in parallel via child workflows.
type ChildFanoutInput struct {
	Items []string `json:"items"`
}

// ChildItemResult holds the result of processing a single item.
type ChildItemResult struct {
	ItemID      string `json:"itemId"`
	Status      string `json:"status"`
	ProcessedBy string `json:"processedBy"`
}

// ChildFanoutResult aggregates all child results.
type ChildFanoutResult struct {
	TotalItems int               `json:"totalItems"`
	Completed  int               `json:"completed"`
	Failed     int               `json:"failed"`
	Results    []ChildItemResult `json:"results"`
}

// ChildFanoutWorkflow fans out N child workflows and waits for all to complete.
// Demonstrates: child workflows, parallel execution, result aggregation, partial failure handling.
func ChildFanoutWorkflow(ctx workflow.Context, input ChildFanoutInput) (*ChildFanoutResult, error) {
	logger := workflow.GetLogger(ctx)
	logger.Info("ChildFanoutWorkflow started", "itemCount", len(input.Items))

	childOpts := workflow.ChildWorkflowOptions{
		WorkflowExecutionTimeout: 2 * time.Minute,
		RetryPolicy: &temporal.RetryPolicy{
			MaximumAttempts: 2,
		},
	}
	ctx = workflow.WithChildOptions(ctx, childOpts)

	// Launch all child workflows in parallel
	var futures []workflow.ChildWorkflowFuture
	for _, item := range input.Items {
		future := workflow.ExecuteChildWorkflow(ctx, ChildProcessItemWorkflow, item)
		futures = append(futures, future)
	}

	// Collect results
	var results []ChildItemResult
	completed := 0
	failed := 0

	for i, future := range futures {
		var result ChildItemResult
		err := future.Get(ctx, &result)
		if err != nil {
			logger.Warn("Child workflow failed", "item", input.Items[i], "error", err)
			results = append(results, ChildItemResult{
				ItemID: input.Items[i],
				Status: fmt.Sprintf("failed: %v", err),
			})
			failed++
		} else {
			results = append(results, result)
			completed++
		}
	}

	logger.Info("ChildFanoutWorkflow completed",
		"total", len(input.Items),
		"completed", completed,
		"failed", failed,
	)

	return &ChildFanoutResult{
		TotalItems: len(input.Items),
		Completed:  completed,
		Failed:     failed,
		Results:    results,
	}, nil
}

// ChildProcessItemWorkflow processes a single item.
// This is the child workflow that gets spawned by ChildFanoutWorkflow.
func ChildProcessItemWorkflow(ctx workflow.Context, itemID string) (*ChildItemResult, error) {
	logger := workflow.GetLogger(ctx)
	logger.Info("ChildProcessItemWorkflow started", "itemId", itemID)

	activityOpts := workflow.ActivityOptions{
		StartToCloseTimeout: 30 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			MaximumAttempts: 2,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, activityOpts)

	var result string
	err := workflow.ExecuteActivity(ctx, activities.ProcessItem, itemID).Get(ctx, &result)
	if err != nil {
		return nil, err
	}

	return &ChildItemResult{
		ItemID:      itemID,
		Status:      "completed",
		ProcessedBy: "go-worker",
	}, nil
}
